### PR TITLE
try to fix issue 25 by ensuring resources are correctly closed

### PR DESCRIPTION
**Pull request by dneise**
The problem with issue #25 was that the `Tables` where not correctly resetting themselves when a File was re-opened. 

This is all connected to the fact that multiple open Files (and multiple open tables) are not yet supposed by the underlying C++/Python wrapper of the protozfits reader. 

At the moment, I simply warn a user who is trying to open multiple files at the same time.

If multiple files happen to be open at the same time, but only the last one is used by the user, then all is fine... just when one tries something like this, one will always receive the same events...

```python
f1 = protozfits.simple.File('foo.zfits')
f2 = protozfits.simple.File('copy_.zfits')
for e1, e2 in zip(f1.Events, f2.Events):
    # somehow compare e1 and e2 or so.
    print(f1.eventNumber, f2.eventNumber)
```

One will receive a warning and the for loop will never terminate, because the Tables will continuously reset themselves...

### CODE REVIEW

File: protozfits/simple.py [hunk 1]
```diff
@@ -22,6 +33,18 @@ def __repr__(self):
...
         )
 
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, tb):
+        self.close()
+
+    def close(self):
+        File.instances = 0
```
Comment (MaxNoe) [inline protozfits/simple.py]: Shouldn't it just subtract 1 instead of setting it to 0?

**Comment by watsonjj** [comment]
Thanks for this PR @dneise.

I wonder if its enough to only warn about the multiple files being opened, someone might not realise the warning is occurring and believe the results from their script are correct. Would it be better to throw an error if someone attempts to read multiple files at once?

**Comment by watsonjj** [comment]
A similar issue happens with the pyhessio library. We handle it by forcibly closing the previous file (inside HESSIOEventSource), so that could be another option. I'd normally add such a thing to ctapipe, however as 2-3 file formats use this package, it would be easier for the package to dictate what to do.

**Comment by dneise** [comment]
Thanks for reviewing @watsonjj 

> I wonder if its enough to only warn about the multiple files being opened, 

I'm not sure either... 

> I'd normally add such a thing to ctapipe ...

I totally agree to handle such peculiarities of the reader in the reader itself. 

---

I guess we should think a bit more about this before merging.

**Comment by dneise** [comment]
One more thing, which I did not yet mention in the README nor anywhere here.

The underlying library does not only not support yet having multiple files open, but also not having multiple Tables open at the same time. So while this here will store the run header in `rh` and the first 3 events in `events`:
```python
f = SimpleFile('filename.fits.fz')
rh = next(f.RunHeader)
events = [next(f.Events) for i in range(3)]
```

Consider this instead: 
Another user might say: I think the run header is somehow important for every event, so I will create tuples of events and run_headers and give these tuples to my algorithms ... 

```python
f = SimpleFile('filename.fits.fz')
rhs_and_events = [ 
    (next(f.RunHeader), next(f.Events))
    for i in range(3)
]
```

This will fail horribly, the user always gets the same event. Why? Well the underlying C++/python wrapper can only have one table open at a time (this causes also the restriction to only one file). But I wanted to create an interface where the user does not see that there are tables opened and closed behind the scenes. So when a user accesses table `RunHeaders` this table is opened (and iteration starts at row 0), when later the user accesses table `Events` this table is opened (and iteration starts at row 0) .. if for some reason the user stops iterating over `Events` after some number of rows ... and wants to look at `RunHeaders` again (or a 3rd Table we might introduce later) then `Events` is closed and `RunHeaders` is opened. After some time our user might want to come back to `Events` and go on with the iteration only to find out that iteration restarted at row 0 ... 

This might be very surprising .. it might even lead to horrible mistakes. On the other hand, each event contains its event id ... so users can *see* that they are not reading the event they assume to be reading... but still it is very surprising. 

Sadly the underlying C++/python wrapper does not allow to jump to a certain event, otherwise we could record at which row we were before "jumping" to another table and upon returning we could "fast forward" to that row and the user would probably get exactly what she assumes to get.

Maybe however resetting to row 0 is exactly what a user expects ... dunno.

**Comment by dneise** [comment]
So at the moment, I tried to go for a "practical" approach. Trying to support what I assume is the usual use case ... jumping back and forth between tables I consider unusual... 

Having multiple files open is not unusual... and it will emit a warning, the warning says having multiple files open is not supported ... but in deed it is just not "fully" supported. 

E.g. one use-case I had in the past in a different project was this:
```
paths = glob('search/path/*.whatever')
files = [reader.open(path) for path in paths]
results = []
for file in files:
    for event in file:
        results.append(analyze(event))
```

And this works ... despite the warning saying that it is not supported ... why does it work? because the user is iterating over everything in series .. not trying to do some kind of parallel access.....

The point is .. I hope in the future the C++/python wrapper will support exactly this behaviour. And If I would throw a lot of exceptions now forcing people to work around that problem in the wrapper, it feels like telling them: "this will never work" ... But it might work in a month .. or two weeks .. or even one week .. or never.

We have to keep in mind .. this file format / reader might turn out to be never used .. so we should not put too much effort into it.

**Pull request merged**